rpcs3/Emu/Cell/SPUThread.cpp

```diff
@@ -3220,9 +3221,73 @@ bool spu_thread::reservation_check(u32 addr, const decltype(rdata)& data)
 	}
 
 	// Ensure data is allocated (HACK: would raise LR event if not)
-	vm::range_lock<false>(range_lock, addr, 128);
+	range_lock->store(u64{128} << 32 | addr);
 
-	const bool res = *range_lock && cmp_rdata(data, vm::_ref<decltype(rdata)>(addr));
+	// Set range_lock first optimistically
+	while (u64 lock_val = vm::g_range_lock)
```

Comment:
We might need in the future in other places such as in vm::try_access, I strongly recommend keeping in function.

Reply:
Keeping locking in function?
Perhaps there are many functions, for different purpose, or you mean one mega-template for all purposes?
This way, it's easier to make changes and optimizations.

Reply:
mega template.

Reply:
specializations such as vn::try_range_lock can be made to allow easier use of these template arguments to the original function.

Reply:
May be hard to do everything at once and have all optimizations...

Reply:
well okay.